services/rpcfilters/api.go

```diff
@@ -1,58 +1,106 @@
 package rpcfilters
 
 import (
+	"context"
 	"errors"
 	"sync"
+	"time"
 
+	ethereum "github.com/ethereum/go-ethereum"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/eth/filters"
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/pborman/uuid"
 )
 
-type filter struct {
-	hashes []common.Hash
-	mu     sync.Mutex
-	done   chan struct{}
-}
-
-// AddHash adds a hash to the filter
-func (f *filter) AddHash(hash common.Hash) {
-	f.mu.Lock()
-	defer f.mu.Unlock()
-	f.hashes = append(f.hashes, hash)
-}
-
-// PopHashes returns all the hashes stored in the filter and clears the filter contents
-func (f *filter) PopHashes() []common.Hash {
-	f.mu.Lock()
-	defer f.mu.Unlock()
-	hashes := f.hashes
-	f.hashes = nil
-	return returnHashes(hashes)
-}
+const (
+	filterLivenessPeriod = 5 * time.Minute
+	logsPeriod           = 10 * time.Second
+	logsQueryTimeout     = 10 * time.Second
+)
 
-func newFilter() *filter {
-	return &filter{
-		done: make(chan struct{}),
-	}
+type filter interface {
+	add(interface{})
+	pop() interface{}
+	stop()
+	deadline() *time.Timer
 }
 
 // PublicAPI represents filter API that is exported to `eth` namespace
 type PublicAPI struct {
-	filters                        map[rpc.ID]*filter
-	filtersMu                      sync.Mutex
+	filtersMu sync.Mutex
+	filters   map[rpc.ID]filter
+
+	filterLivenessLoop   time.Duration
+	filterLivenessPeriod time.Duration
+
+	client func() ContextCaller
+
 	latestBlockChangedEvent        *latestBlockChangedEvent
 	transactionSentToUpstreamEvent *transactionSentToUpstreamEvent
 }
 
 // NewPublicAPI returns a reference to the PublicAPI object
-func NewPublicAPI(latestBlockChangedEvent *latestBlockChangedEvent,
-	transactionSentToUpstreamEvent *transactionSentToUpstreamEvent) *PublicAPI {
-	return &PublicAPI{
-		filters:                        make(map[rpc.ID]*filter),
-		latestBlockChangedEvent:        latestBlockChangedEvent,
-		transactionSentToUpstreamEvent: transactionSentToUpstreamEvent,
+func NewPublicAPI(s *Service) *PublicAPI {
+	api := &PublicAPI{
+		filters:                        make(map[rpc.ID]filter),
+		latestBlockChangedEvent:        s.latestBlockChangedEvent,
+		transactionSentToUpstreamEvent: s.transactionSentToUpstreamEvent,
+		client:               func() ContextCaller { return s.rpc.RPCClient() },
+		filterLivenessLoop:   filterLivenessPeriod,
+		filterLivenessPeriod: filterLivenessPeriod + 10*time.Second,
 	}
+	go func() {
+		api.timeoutLoop(s.quit)
+	}()
+	return api
+}
+
+func (api *PublicAPI) timeoutLoop(quit chan struct{}) {
+	for {
+		select {
+		case <-quit:
+			return
+		default:
+		}
+		time.Sleep(api.filterLivenessLoop)
```

Comment:
Is it ok to proceed with the logic after `time.Sleep` if `quit` channel is closed? Should it be:
```go
for {
  select {
  case <-quit:
    return
  case <-time.After(api.filterLivenessLoop):
  }
}
```